### PR TITLE
Support generalize data to be store in MDL

### PR DESCRIPTION
**Pull request by mykytaccode0**
- https://freedcamp.com/view/2956207/tasks/40335913
- Support flow art like, follow/unfollow, friend

### CODE REVIEW

File: common/service/userdata/request.go [hunk 1]
```diff
@@ -96,4 +174,182 @@ type SuperNodeReply struct {
...
+	// PastelID of the actor that sign this data
+	PastelID string
+	// Command the the predefine database operation name
+	Command string
+	// Data is the general structure accept many kind of metric, but it need to be match with Command
+	Data []byte
+}
+
+// ArtInfo is representation of an art
+type ArtInfo struct {
```
Comment (leonidcode20) [inline common/service/userdata/request.go:194]: I think to align with other flows "art" should be "nft"

File: common/service/userdata/request.go [hunk 2]
```diff
@@ -1,5 +1,77 @@
 package userdata
 
+import "time"
+
+const (
```
Comment (leonidcode20) [inline common/service/userdata/request.go:5]: Could we use just one term "art" or "nft"

File: common/service/userdata/request.go [hunk 3]
```diff
@@ -96,4 +174,182 @@ type SuperNodeReply struct {
...
+	ArtistPastelID        string  `mapstructure:"artist_pastel_id" json:"artist_pastel_id"`
+	Copies                int64   `mapstructure:"copies" json:"copies"`
+	CreatedTimestamp      int64   `mapstructure:"created_timestamp" json:"created_timestamp"`
+	GreenNft              bool    `mapstructure:"green_nft" json:"green_nft"`
+	RarenessScore         float64 `mapstructure:"rareness_score" json:"rareness_score"`
+	RoyaltyRatePercentage float64 `mapstructure:"royalty_rate_percentage" json:"royalty_rate_percentage"`
+}
+
+// ArtInstanceInfo is representation of an art instance
+type ArtInstanceInfo struct {
```
Comment (leonidcode20) [inline common/service/userdata/request.go:205]: ditto

File: metadb/database/safe_string_query.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+func processInputString(s string) string {
+	if s == "" {
+		return "NULL"
+	}
+	return processEscapeString(s)
+}
+
+func safeStringQueryStruct(ps interface{}) interface{} {
+	v := reflect.ValueOf(ps).Elem() // Elem() dereferences pointer
```
Comment (leonidcode20) [inline metadb/database/safe_string_query.go:21]: Should check if ps is `Struct` ortherwise will panic.

File: metadb/database/service.go [hunk 5]
```diff
@@ -16,41 +15,58 @@ import (
...
+	getInstanceInfoTemplate          = "get_instance_info"
+	getTopSNActivitiesTemplate       = "get_top_sn_activities"
+	getUserBriefInfoTemplate         = "get_user_brief_info"
+	highestSalePriceByUserTemplate   = "highest_sale_price_by_user"
+	newArtAuctionTemplate            = "new_art_auction"
+	nftCopiesExistTemplate           = "nft_copies_exist"
+	nftCreatedByArtistTemplate       = "nft_created_by_artist"
+	nftForSaleByArtistTemplate       = "nft_for_sale_by_artist"
+	nftOwnedByUserTemplate           = "nft_owned_by_user"
+	nftSoldByArtIDTemplate           = "nft_sold_by_art_id"
```
Comment (leonidcode20) [inline metadb/database/service.go:45]: I think "nft" and "art" are exclusive term. Better is "nft_sold_by_id".

File: metadb/database/service.go [hunk 6]
```diff
@@ -95,6 +456,9 @@ func (db *Ops) ReadUserData(ctx context.Context, artistPastelID string) (userdat
 	if err != nil {
 		return userdata.ProcessRequest{}, errors.Errorf("error while querying db: %w", err)
 	}
+	if queryResult.Err != nil {
+		return userdata.ProcessRequest{}, errors.Errorf("error while querying db: %w", queryResult.Err)
+	}
 
 	nrows := queryResult.NumRows()
```
Comment (leonidcode20) [inline metadb/database/service.go:463]: do we need to "Close" the queryResult of something similar ?

File: walletnode/api/services/helper.go [hunk 7]
```diff
@@ -301,3 +302,19 @@ func toUserSpecifiedData(req *userdata.ProcessRequest) *userdatas.UserSpecifiedD
 
 	return result
 }
+
+func toRelationshipInfoArray(items []userdata.UserRelationshipItem) []*userdatas.UserRelationshipInfo {
+	if items == nil {
+		return nil
+	}
+	result := make([]*userdatas.UserRelationshipInfo, 0)
```
Comment (leonidcode20) [inline walletnode/api/services/helper.go:310]: Better `make([]*userdatas.UserRelationshipInfo, len(items)`

File: walletnode/api/design/userdatas.go [hunk 8]
```diff
@@ -286,3 +396,150 @@ var GetUserdataParams = func() {
...
+	Attribute("user_pastel_id", String, func() {
+		Description("User's PastelID")
+		MinLength(86)
+		MaxLength(86)
+		Pattern(`^[a-zA-Z0-9]+$`)
+		Example("jXYJud3rmrR1Sk2scvR47N4E4J5Vv48uCC6se2nzHrBRdjaKj3ybPoi1Y2VVoRqi1GnQrYKjSxQAC7NBtvtEdS")
+	})
+
+	Attribute("art_pastel_id", String, func() {
+		Description("Art's PastelID")
```
Comment (leonidcode20) [inline walletnode/api/design/userdatas.go:432]: Can we reuse the `pastelid` in other file or consolidate to one type ?

File: supernode/services/userdataprocess/task.go [hunk 9]
```diff
@@ -511,6 +512,55 @@ func (task *Task) ConnectToLeader(ctx context.Context, extAddress string) error
...
+		}
+		if err := node.connect(ctx); err != nil {
+			log.WithContext(ctx).Debugf("Fail to connect to node %s with error: %s", mn.ExtAddress, err.Error())
+			continue
+		}
+		// Send data to the highest ranked SN that we just connect
+		_, err = node.StoreMetric(ctx, metric)
+		if err != nil {
+			// If there is issue then we return
+			return err
```
Comment (leonidcode20) [inline supernode/services/userdataprocess/task.go:555]: `return errors.Errorf("failed to store metric: %w", err)`

File: proto/walletnode/process_userdata_walletnode.proto [hunk 10]
```diff
@@ -80,6 +80,10 @@ message UserdataRequest {
   string userdata_hash = 14;
   // Signature of the userdata_hash
   string signature = 15;
+  // Command is database operation that will apply to the data
+	string command = 16;
```
Comment (leonidcode20) [inline proto/walletnode/process_userdata_walletnode.proto:84]: Fix alignment please

File: proto/supernode/process_userdata_supernode.proto [hunk 11]
```diff
@@ -43,6 +46,8 @@ message SuperNodeReply {
   int32 response_code = 1;
   // The detail of why result is success/fail, depend on response_code
   string detail = 2;
+  // Data is the walletnode metric that MetadataLayer return
```
Comment (leonidcode20) [inline proto/supernode/process_userdata_supernode.proto:49]: Fix alignment please.

**Comment by a-ok123** [comment]
this is way outdated now

**Pull request closed without merging**